internal/db/query.go

```diff
@@ -291,6 +305,62 @@ func GetTars(ctx context.Context, tx pgx.Tx, project int64, vrange VersionRange,
 	}, nil
 }
 
+func latestCacheTarsQuery() string {
+	return `
+		WITH latest_cache_version AS (
+			SELECT version, hashes
+			FROM dl.cache_versions
+			ORDER BY version DESC LIMIT 1
+		),
+		version_hash AS (
+			SELECT version, unnest(hashes) AS hash
+			FROM latest_cache_version
+		)
+		SELECT version, (version_hash.hash).h1, (version_hash.hash).h2, bytes
+		FROM version_hash
+		JOIN dl.contents
+		ON version_hash.hash = dl.contents.hash
+	`
+}
+
+type cacheTarStream func() (int64, []byte, []byte, error)
+
+func GetCacheTars(ctx context.Context, tx pgx.Tx) (cacheTarStream, error) {
+	rows, err := tx.Query(ctx, latestCacheTarsQuery())
+	if err != nil {
+		return nil, fmt.Errorf("GetCachedTars query: %w", err)
+	}
+
+	var version int64
+
+	stream := func() (int64, []byte, []byte, error) {
+		if !rows.Next() {
+			return 0, nil, nil, io.EOF
+		}
+		var rowVersion int64
+		var h1, h2 []byte
+		var encoded []byte
+
+		err := rows.Scan(&rowVersion, &h1, &h2, &encoded)
+		if err != nil {
+			return 0, nil, nil, fmt.Errorf("GetCacheTars scan: %w", err)
+		}
+		if version != 0 && rowVersion != version {
+			return 0, nil, nil, fmt.Errorf("GetCacheTars rowVersion mismatch: %d vs %d", version, rowVersion)
+		}
+
+		version = rowVersion
+
+		hash := make([]byte, 32)
+		copy(hash[0:16], h1)
+		copy(hash[16:], h2)
```

Comment:
Another case where the `Hash` type will add better type info to this function and allow you to not have to repeat this code.